gh-67022: Document bytes/str inconsistency in email.header.decode_header() and suggest email.headerregistry.HeaderRegistry as a sane alternative

dlenski opened the pull request:
This function's possible return types have been surprising and error-prone
for the entirety of its Python 3.x history. It can return either:

    typing.List[typing.Tuple[str, None]], of length exactly 1
    or typing.List[typing.Tuple[bytes, typing.Optional[str]]]

This function can't be rewritten to be more consistent in a backwards-compatible way, because some users of this function depend on the existing return type(s).

This PR addresses the inconsistencies:

1. as suggested by @JelleZijlstra in https://github.com/python/cpython/issues/67022#issuecomment-1093668884:

   > we should document the surprising return type at https://docs.python.org/3.10/library/email.header.html.
3. by suggesting `email.headerregistry.HeaderRegistry` as a replacement, per https://github.com/python/cpython/pull/92900/#discussion_r1112472177

Example of the old/inconsistent (`decode_header`) vs. modern/sane approaches:

```py
>>> from email import decode_header
>>> from email.headerregistry import HeaderRegistry
>>>
>>> # decode_header exposes differences in sub-encodings AND a str/bytes inconsistency in return type:
>>> print(decode_header('hello foo bar')
 [('hello foo bar', None)])
>>> print(decode_header('hello =?utf-8?B?ZsOzbw==?= bar'))
[(b'hello ', None), (b'f\xc3\xb3o', 'utf-8'), (b' bar', None)]
>>> print(decode_header('=?iso-8859-1?q?hello_f=F3o_bar?='))
[(b'hello f\xf3o bar', 'iso-8859-1')]
>>>
>>> # HeaderRegistry has a sane and consistent interface:
>>> decoder = HeaderRegistry()
>>> decoder('Subject', 'hello foo bar')
'hello foo bar'
>>> decoder('Subject', 'hello =?utf-8?B?ZsOzbw==?= bar')
'hello fóo bar'
>>> decoder('Subject', '=?iso-8859-1?q?hello_f=F3o_bar?=')
'hello fóo bar'
```

_(Closes #30548 and replaces it.)_

<!-- gh-issue-number: gh-67022 -->
* Issue: gh-67022
<!-- /gh-issue-number -->

dlenski committed:
gh-67022: Document bytes/str inconsistency in email.header.decode_header()

This function's possible return types have been surprising and error-prone
for the entirety of its Python 3.x history. It can return either:

1. `typing.List[typing.Tuple[bytes, typing.Optional[str]]]` of length >1
2. or `typing.List[typing.Tuple[str, None]]`, of length exactly 1

This means that any user of this function must be prepared to accept either
`bytes` or `str` for the first member of the 2-tuples it returns, which is a
very surprising behavior in Python 3.x, particularly given that the second
member of the tuple is supposed to represent the charset/encoding of the
first member.

This patch documents the behavior of this function, and adds test cases
to demonstrate it.

As discussed in bpo-22833, this cannot be changed in a backwards-compatible
way, and some users of this function depend precisely on the existing
behavior.

warsaw commented on Doc/library/email.header.rst:
```diff
@@ -173,21 +173,20 @@ Here is the :class:`Header` class description:
-   Decode a message header value without converting the character set. The header
-   value is in *header*.
+   Decode a message header value to a Unicode string, including handling
+   portions encoded according to :rfc:`2047`.
```
What happens if this function is called on a malformed encoded string?

dlenski replied:
The underlying `decode_header` is extremely tolerant and will treat almost any mangled string as if it's simply a "plain" unencoded string…

```
print(email.header.decode_header('=?-fo!!?o-bar=q=this is some text=?'))
[('=?-fo!!?o-bar=q=this is some text=?', None)]
```

warsaw commented on Doc/library/email.header.rst:
```diff
@@ -203,3 +202,34 @@ The :mod:`email.header` module also provides the following convenient functions.
+   .. note::
+
+      This function exists for for backwards compatibility only. For
+      new code we recommend using :mod:`email.header.decode_header_to_string`.
```
How about adding a link to the non-legacy API, or an example using that newer API?

dlenski replied:
Do you mean the non-legacy API as described in https://docs.python.org/3/library/email.html, e.g. `email.parser`?

To my knowledge, there is not any function/method in that API which can be straightforwardly used instead of `email.header.decode_header`.

bitdancer replied:
```
>>> from email.headerregistry import HeaderRegistry
>>> decoder = HeaderRegistry()
>>> decoder('To', '=?utf-8?q?M=C3=A4x?= <foo@bar.com>')
'Mäx <foo@bar.com>'
>>> decoder('To', '=?utf-8?q?M=C3=A4x?= <foo@bar.com>').addresses
(Address(display_name='Mäx', username='foo', domain='bar.com'),)
```

You really don't want to use the legacy decode_header.  It has many bugs that the new API fixes.

bitdancer replied:
Obviously this needs to be better documented...

dlenski replied:
Revisiting this after a long while. My original motivation for this was to robustly decode an application-specific `content-type`-like header in a proxying application… I've run into it again in a similar bit of HTTP middleware. 😅

So is it fair to say that the _correct_ way to handle this now is with a method like the following?

```py
from email.headerregistry import HeaderRegistry

_decoder = HeaderRegistry()
def decode_http_header(header: str, value: str):
    return _decoder(str, value)
```

dlenski replied:
I've adopted `HeaderRegistry` as a suggested replacement instead.

bitdancer replied:
Yes, that works.

An alternative would be:
```
from email.policy import HTTP

decode_http_header = HTTP.header_factory
```

but I don't know that it is better.  It *might* be slightly more future proof, but no guarantees ;)  It has the disadvantage that it is using a module global that could get modified (legitimately) from other code; whether that is a concern or not depends on the application's structure.

warsaw commented on Lib/email/header.py:
```diff
@@ -152,6 +158,23 @@ def decode_header(header):
+    or it may be a Header object; in the latter case, this is equivalent to
+    str(header).
+
+    An email.errors.HeaderParseError may be raised when certain decoding error
```
This comment should be lifted into the stdlib documentation.

dlenski replied:
Ack, will do.

warsaw commented on Misc/NEWS.d/next/Library/2022-01-11-21-40-14.bpo-22833.WB-JWw.rst:
```diff
@@ -0,0 +1,4 @@
+The inconsistent return types of :func:`email.header.decode_header` are now documented.
+
+:func:`email.header.decode_header_to_string` is provided as a less error-prone and
+more straightforward replacement for it.
```
I suggest "alternative" instead of "replacement".

dlenski replied:
Ack, will do.

warsaw commented on Lib/email/header.py:
```diff
@@ -152,6 +158,23 @@ def decode_header(header):
 
 
+
+def decode_header_to_string(header):
```
I'm not seeing any tests of this function added in the PR.  Am I missing that (I'm possibly not navigating the iPadOS GitHub client correctly yet).

dlenski replied:
No, you're not missing that.

The implementation is so simple that I didn't think they were necessary, but I could add some reasonably-orthogonal unit tests here.

warsaw reviewed (changes requested):
In general, I think this would help users of the legacy API, although I think we should also steer people to the new API. What does @bitdancer think?

dlenski committed:
Add warnings about obsolescence of 'email.header.decode_header' and 'email.header.make_header' functions.

Recommend use of `email.headerregistry.HeaderRegistry` instead, as suggested
in https://github.com/python/cpython/pull/92900#discussion_r1112472177

bitdancer commented on Misc/NEWS.d/next/Library/2022-01-11-21-40-14.bpo-22833.WB-JWw.rst:
```diff
@@ -0,0 +1,4 @@
+The inconsistent return types of :func:`email.header.decode_header` are now documented,
```
I don't think we do news items for doc-only PRs.  I've added the skip-news tag. (One of the triagers can correct me if I'm wrong:)

dlenski replied:
Thanks. Done in 4b6143ec

bitdancer reviewed (changes requested):
Looks good to me except for the news item (assuming I'm right about that).

dlenski committed:
Removed NEWS item

Per https://github.com/python/cpython/pull/92900#discussion_r2141163470, not
wanted for doc-only PRs.

dlenski committed:
Merge branch 'main' into gh60722

bitdancer reviewed (approved):
LGTM
